File: trees/my_calendar_II.rb
class Node
  attr_accessor :s, :e, :left, :right, :single_overlap

  def initialize(s, e)
    @s = s
    @e = e
    @left = nil
    @right = nil
    @single_overlap = false
  end
end

class MyCalendarTwo
  def initialize()
    @root = nil
  end

  # @param {TreeNode} root
  # @return Boolean
  def book(s, e)
    return false if !insertable?(s, e, @root)

    @root = insert(s, e, @root)
    true
  end

  def insertable?(s, e, root)
    return true if root.nil?
    return true if s >= e

    if e <= root.s
      insertable?(s, e, root.left)
    elsif s >= root.e
      insertable?(s, e, root.right)
    else # Overlap
      if root.single_overlap
        false
      elsif s >= root.s && e <= root.e
        true
      else
        insertable?(s, root.s, root.left) && insertable?(root.e, e, root.right)
      end
    end
  end

  private
  # O(logH) solution, where H is the height of BT - we can improve further if we use balanced binary trees
  def insert(s, e, root)
    return root = Node.new(s, e) if root.nil?
    return root if s >= e

    if s >= root.e
      root.right = insert(s, e, root.right)
    elsif e <= root.s
      root.left = insert(s, e, root.left)
    else # Overlap
      root.single_overlap = true
      a = [root.s, s].min
      b = [root.s, s].max
      c = [root.e, e].min
      d = [root.e, e].max

      root.s, root.e = b, c
      root.left, root.right = insert(a, b, root.left), insert(c, d, root.right)
    end

    root
  end
end


require 'test/unit'
extend Test::Unit::Assertions

calendar = MyCalendarTwo.new()
assert_equal(calendar.book(10, 20), true)
assert_equal(calendar.book(50, 60), true)
assert_equal(calendar.book(10, 40), true)
assert_equal(calendar.book(5, 15), false)
assert_equal(calendar.book(5, 10), true)
assert_equal(calendar.book(25, 55), true)

File: trees/my_calendar_II.py
class MyCalendarTwo:
	
	def __init__(self):
		self.root = None

	def book(self, start, end):
		"""
		:type start: int
		:type end: int
		:rtype: bool
		"""
		if not self.is_insertable(start, end, self.root):
			return False

		self.root = self.insert(start, end, self.root)
		return True

	def is_insertable(self, start, end, root):
		if not root:
			return True

		if start >= end:
			return True

		if end <= root.start:
			return self.is_insertable(start, end, root.left)

		elif start >= root.end:
			return self.is_insertable(start, end, root.right)

		else: #overlap
			if root.single_overlap:
				return False
			elif start >= root.start and end <= root.end:
				return True
			else:
				return self.is_insertable(start, root.start, root.left) and self.is_insertable(root.end, end, root.right)

	def insert(self, start, end, root):
		if not root:
			root = Node(start, end)
			return root

		if start >= end:
			return root

		if start >= root.end:
			root.right = self.insert(start, end, root.right)

		elif end <= root.start:
			root.left = self.insert(start, end, root.left)

		else:
			root.single_overlap = True
			a = min(root.start, start)
			b = max(root.start, start)
			c = min(root.end, end)
			d = max(root.end, end)
			root.start, root.end = b, c
			root.left, root.right = self.insert(a, b, root.left), self.insert(c, d, root.right)

		return root

class Node:
	"""
	:type left: Node
	:type right: Node
	:type single_overlap: bool
	"""
	def __init__(self, start, end):
		"""
		:type start: int
		:type end: int
		"""
		self.start = start
		self.end = end
		self.left = None
		self.right = None
		self.single_overlap = False

# Your MyCalendarTwo object will be instantiated and called as such:
# obj = MyCalendarTwo()
# param_1 = obj.book(start,end)
